llvm/utils/lit/lit/builtin_commands/cat.py

```diff
@@ -1,29 +1,45 @@
 import getopt
 import sys
+from dataclasses import dataclass, fields
 
 try:
     from StringIO import StringIO
 except ImportError:
     from io import StringIO
 
 
-def convertToCaretAndMNotation(data):
+@dataclass
+class Options:
+    show_ends: bool
+    show_nonprinting: bool
+
+
+def convertTextNotation(data, options):
     newdata = StringIO()
     if isinstance(data, str):
         data = bytearray(data.encode())
 
+
     for intval in data:
-        if intval == 9 or intval == 10:
-            newdata.write(chr(intval))
-            continue
-        if intval > 127:
-            intval = intval - 128
-            newdata.write("M-")
-        if intval < 32:
-            newdata.write("^")
-            newdata.write(chr(intval + 64))
-        elif intval == 127:
-            newdata.write("^?")
+        if options.show_ends:
+            if intval == 10:
```

Comment:
nit: you can combine these and drop a level of nesting. Could also be `if/else if/else` but that isn't too important since there's a continue.

Reply:
done